test/integration/protocol_integration_test.cc

```diff
@@ -4921,4 +4921,64 @@ TEST_P(DownstreamProtocolIntegrationTest, UnknownPseudoHeader) {
   }
 }
 
+// This test performs DNS refreshing very fast also keep changing DNS results.
+// It is trying to reproduce a R/W race for logical_dns_cluster.
+class LogicalDnsReadWriteRaceTest : public ProtocolIntegrationTest {
+public:
+  LogicalDnsReadWriteRaceTest() : registered_dns_factory_(dns_resolver_factory_) {}
+
+  NiceMock<Network::MockDnsResolverFactory> dns_resolver_factory_;
+  Registry::InjectFactory<Network::DnsResolverFactory> registered_dns_factory_;
+};
+
+INSTANTIATE_TEST_SUITE_P(
+    Protocols, LogicalDnsReadWriteRaceTest,
+    testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParamsWithoutHTTP3()),
+    HttpProtocolIntegrationTest::protocolTestParamsToString);
+
+TEST_P(LogicalDnsReadWriteRaceTest, FastChangingDnsResult) {
+  if (version_ == Network::Address::IpVersion::v4) {
+    return;
+  }
+  std::shared_ptr<Network::MockDnsResolver> dns_resolver(new Network::MockDnsResolver());
+  EXPECT_CALL(dns_resolver_factory_, createDnsResolver(_, _, _))
+      .WillRepeatedly(testing::Return(dns_resolver));
+  EXPECT_CALL(*dns_resolver, resolve(_, _, _))
+      .WillRepeatedly(
+          Invoke([&](const std::string&, Network::DnsLookupFamily,
+                     Network::DnsResolver::ResolveCb dns_callback) -> Network::ActiveDnsQuery* {
+            // Keep changing the DNS response address list.
+            static int i = 1;
+            dns_callback(Network::DnsResolver::ResolutionStatus::Success,
+                         TestUtility::makeDnsResponse({"::1", ("127.0.0." + std::to_string(i)),
+                                                       ("127.0.0." + std::to_string(i + 1))}));
+            i = (i+1) % 128;
+            return nullptr;
+          }));
+  if (use_universal_header_validator_) {
```

Comment:
what does the header validator have to do with anything?

Also if you are going to early exit, shouldn't you do it at the top?

Reply:
UHV should not affect the test, maybe cut/paste from another test?

Reply:
Yeah,  other than the DNS resolver mocking and DNS cluster configuration, the rest of test code is same as the above LogicalDns test: https://github.com/envoyproxy/envoy/blob/d82b5a02befbff3b2119b309e865f7b44d207907/test/integration/protocol_integration_test.cc#L73.